ChemistryLib/PhreeqcKernel.cpp

```diff
@@ -0,0 +1,257 @@
+            master_bsearch(transport_process_variable.c_str());
+
+        _process_id_to_master_map[transport_process_id] = master_species;
+    }
+}
+
+void PhreeqcKernel::reinitializeRates()
+{
+    count_rates = _reaction_rates.size();
+    rates = (struct rate*)realloc(
+        rates, (std::size_t)(count_rates) * sizeof(struct rate));
+    int rate_id = 0;
+    for (auto const& reaction_rate : _reaction_rates)
+    {
+        rates[rate_id].name = reaction_rate.kinetic_reactant.data();
+
+        // Commands' strings are freed by Phreeqc dtor.
+        rates[rate_id].commands = static_cast<char*>(
+            malloc(sizeof(char) * reaction_rate.commands().size() + 1));
+        if (rates[rate_id].commands == nullptr)
+            OGS_FATAL("Could not allocate memory for rate[%d] commands.",
+                      rate_id);
+        reaction_rate.commands().copy(rates[rate_id].commands,
+                                      std::string::npos);
+        rates[rate_id].commands[reaction_rate.commands().size()] = '\0';
+
+        rates[rate_id].new_def = 1;
+        rates[rate_id].linebase = nullptr;
+        rates[rate_id].varbase = nullptr;
+        rates[rate_id].loopbase = nullptr;
+
+        ++rate_id;
+    };
+}
+
+void PhreeqcKernel::doWaterChemistryCalculation(
+    std::vector<GlobalVector*>& process_solutions, double const dt)
+{
+    setAqueousSolutions(process_solutions);
+    setTimeStepSize(dt);
+
+    execute(process_solutions);
+}
+
+void PhreeqcKernel::setAqueousSolutions(
+    std::vector<GlobalVector*> const& process_solutions)
+{
+    // Loop over chemical systems
+    auto const num_chemical_systems = process_solutions[0]->size();
+    for (auto chemical_system_id = 0; chemical_system_id < num_chemical_systems;
+         ++chemical_system_id)
+    {
+        auto& aqueous_solution = Rxn_solution_map[chemical_system_id];
+
+        auto initial_aqueous_solution =
+            getOrCreateInitialAqueousSolution(aqueous_solution);
+
+        auto& components = initial_aqueous_solution->Get_comps();
+        // Loop over transport process id map to retrieve component
+        // concentrations from process solutions
+        for (auto const& map_pair : _process_id_to_master_map)
+        {
+            auto const& transport_process_id = map_pair.first;
+            auto const& master_species = map_pair.second;
+
+            auto& transport_process_solution =
+                process_solutions[transport_process_id];
+
+            auto& element_name = master_species->elt->name;
+            if (isHydrogen(element_name))
+            {
+                // Set pH value by hydrogen concentration.
+                double const pH = -std::log10(
+                    transport_process_solution->get(chemical_system_id));
+                aqueous_solution.Set_ph(pH);
+            }
+            else
+            {
+                // Set component concentrations.
+                auto const& concentration =
```

Comment:
if this is a `double`, drop the reference. Store simple and short types by value.

Reply:
Fixed.